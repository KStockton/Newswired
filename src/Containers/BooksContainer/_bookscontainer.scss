
.book-wrapper {
width: 1000px;
display: grid;
grid-template-columns: repeat(15, 1fr);
grid-auto-flow: row;
overflow: scroll;
width: 100vw;
}




