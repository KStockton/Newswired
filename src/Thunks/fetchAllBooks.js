import * as action from '../actions/index'



export const fetchAllBooks = (url) => {

  return async (dispatch) => {
    try {
      dispatch(action.isLoading(true))
      const response = await fetch(url)
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const books = await response.json()
      dispatch(action.isLoading(false));
      dispatch(action.getAllBooks(books))
    } catch (error) {
      dispatch(action.hasErrored(error.message))
    }
  }
}