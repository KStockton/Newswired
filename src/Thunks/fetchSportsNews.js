import * as action from '../actions/index';


export const fetchSportsNews = (url) => {

  return async (dispatch) => {
    try {
      dispatch(action.isLoading(true))
      const response = await fetch(url)
      if(!response.ok) {
      throw Error(response.sendText)
    }
      const data = await response.json()
      dispatch(action.getSportsNews(data))
      dispatch(action.isLoading(false))
    } catch(error) {
      dispatch(action.hasErrored(error.message, 'sportsNews'))
    }
  }
}