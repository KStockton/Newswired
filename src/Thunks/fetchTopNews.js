import * as action from '../actions/index';


export const fetchTopNews = (url) => {

  return async (dispatch) => {
    try {
      dispatch(action.isLoading(true))
      const response = await fetch(url)
      if(!response.ok) {
      throw Error(response.sendText)
    }
      const data = await response.json()
      dispatch(action.getTopNews(data))
      dispatch(action.isLoading(false))
    } catch(error) {
      dispatch(action.hasErrored(error.message, 'topNews'))
    }
  }
}

