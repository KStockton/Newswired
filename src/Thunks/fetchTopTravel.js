import * as action from '../actions/index';


export const fetchTopTravel = (url) => {
console.log('url', url)
  return async (dispatch) => {
    try {
      dispatch(action.isLoading(true))
      const response = await fetch(url)
      if(!response.ok) {
      throw Error(response.sendText)
    }
      const data = await response.json()
      console.log('data', data)
      dispatch(action.getTopTravel(data))
      dispatch(action.isLoading(false))
    } catch(error) {
      dispatch(action.hasErrored(error.message))
    }
  }
}

