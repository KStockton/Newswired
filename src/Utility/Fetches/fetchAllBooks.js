import { isLoading } from '../../actions/index'



export const fetchAllBooks = () => {
const url = `https://api.nytimes.com/svc/books/v3/lists/current/hardcover-fiction.json?api-key=${API_KEY}`

return async (dispatch) => {
  try {
    dispatch(isLoading(true))
    const response = await fetch(url)
    if(!response.ok) {
      throw new Error(response.statusText)
    }
  } catch (error)
}

}