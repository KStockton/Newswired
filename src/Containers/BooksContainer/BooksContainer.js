import React, { Component } from 'react';
import Book from '../../Components/Book/Book';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchAllBooks } from '../../Thunks/fetchAllBooks';
import {NYT_KEY} from '../../Utility/Config/Key';
import Error from '../../Components/Error/Error'
const shortid = require('shortid');

 class BooksContainer extends Component {
  
async componentDidMount() {
  console.log(this.props.allBooks.length)
  if(this.props.allBooks.length == 0){
    const bookUrl = `https://api.nytimes.com/svc/books/v3/lists/current/hardcover-fiction.json?&api-key=${NYT_KEY}`
     this.props.fetchAllBooks(bookUrl)
     const historyUrl = `https://api.nytimes.com/svc/books/v3/lists/current/hardcover-nonfiction.json?&api-key=${NYT_KEY}`
     this.props.fetchAllBooks(historyUrl)
  }
}



  displayBooks = () => {
  if(this.props.allBooks.length > 0){
    return this.props.allBooks.map(book => {
      return (<Book {...book} key={shortid.generate()} />
      )
    })
  }
  }

  render() {
    const errorMsg = <h2>{this.props.error}</h2>
    const displayBooks = this.displayBooks()    
    return (
      <section>
        <h2>Top NewYork Times Books</h2>
      <div className='BookContainer-wrapper'>
        {displayBooks}
        <p className="error-text">{this.props.error && errorMsg }</p>
      </div>
      </section>
    )
  }
}

BooksContainer.propTypes = {
  allBooks: PropTypes.array
}

export const mapStateToProps = (state) => ({
  allBooks: state.allBooks,
  error: state.error
});

export const mapDispatchToProps = (dispatch) => ({
  fetchAllBooks: (url) => dispatch(fetchAllBooks(url))
})


export default connect(mapStateToProps, mapDispatchToProps)(BooksContainer)
