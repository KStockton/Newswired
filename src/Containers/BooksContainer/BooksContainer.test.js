import React from 'react';
import {BooksContainer, mapStateToProps, mapDispatchToProps} from './BooksContainer';
import { shallow  } from 'enzyme';
import { fetchAllBooks } from './../../Thunks/fetchAllBooks'
jest.mock('../../Thunks/fetchAllBooks')



describe('BookContainer', () => {
  let wrapper;
  let mockFetchFunction = jest.fn()
  let mockBooks = [{title:'Golden'},{title:'Xavier'}]
  let mockError = 'something went wrong'

  beforeEach(() => {
    
    wrapper = shallow(<BooksContainer 
      allBooks={mockBooks}
      error={mockError}
      fetchAllBooks={mockFetchFunction}

      />)
  })
  xit('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  xit('should match the snapshot the books are passed as props', () => {
    wrapper.setProps({allBooks:  [{title: 'Press, Press, Press, Press, Press'}]})
    expect(wrapper).toMatchSnapshot()
  });
  xit('should match the snapshot when there is an error message',() => {
    wrapper.setProps({ allBooks: []}) 
    wrapper.setProps({ error: ''})
    expect(wrapper).toMatchSnapshot()
  });
  it('should call display books when rendered', () => {
    wrapper.instance().displayBooks = jest.fn()
    wrapper.update()
    wrapper.instance().render()
    expect(wrapper.instance().displayBooks).toHaveBeenCalled()
  });
  it('should render books when all books props are passed',() => {
   let books = wrapper.instance().displayBooks().length 
    expect(books).toBe(2)
  });
  it('should not call fetchAllBooks when the allBooks length is greater than 0', () => {  
    let mockBooks = [{title:'Golden'},{title:'Xavier'}]
    wrapper = shallow(<BooksContainer
      allBooks={mockBooks}
      error={mockError}
      />)
    wrapper.instance().componentDidMount()
    expect(fetchAllBooks).not.toHaveBeenCalled()
  })
  it('should call fetchAllBooks when the allBooks length is 0', () => {  
    let mockBooks = []
    wrapper = shallow(<BooksContainer
      allBooks={mockBooks}
      error={mockError}
      fetchAllBooks={mockFetchFunction}
      />)
    wrapper.instance().componentDidMount()
    expect(mockFetchFunction).toHaveBeenCalled()
  });

});
  describe('MSTP',() =>{
    it('should map state from mockstate', () => {
      const mockState ={ allBooks :[{title:'Golden'},{title:'Xavier'}], error: ''}
      const result = mapStateToProps(mockState)
      expect(result).toEqual(mockState)
    });
  });
  describe('MDTP', () => {
    let mockDispatch
    let mappedProps
    
    beforeEach(() => {
        mockDispatch = jest.fn()
        mappedProps = mapDispatchToProps(mockDispatch)
    });

    it('should call fetchAllBooks with the correct params',() => {
      const mockbookUrl = 'www.cheers.com'
      const actionToDispatch = fetchAllBooks(mockbookUrl)
      mappedProps.fetchAllBooks(mockbookUrl)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })

