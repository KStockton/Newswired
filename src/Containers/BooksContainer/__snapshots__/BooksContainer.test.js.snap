// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookContainer should match the snapshot 1`] = `
<section>
  <h2
    className="book-title-header"
  >
     New York Times Best Sellers
  </h2>
  <div
    className="book-wrapper"
  >
    <Connect(Book)
      key="Ga3O1Z3rr"
      title="Golden"
    />
    <Connect(Book)
      key="-nEz7ZrzGH"
      title="Xavier"
    />
  </div>
</section>
`;

exports[`BookContainer should match the snapshot the books are passed as props 1`] = `
<section>
  <h2
    className="book-title-header"
  >
     New York Times Best Sellers
  </h2>
  <div
    className="book-wrapper"
  >
    <Connect(Book)
      key="iw61HL5Z_z"
      title="Press, Press, Press, Press, Press"
    />
  </div>
</section>
`;

exports[`BookContainer should match the snapshot when there is an error message 1`] = `
<section>
  <h2
    className="book-title-header"
  >
     New York Times Best Sellers
  </h2>
  <div
    className="book-wrapper"
  >
    <Loading />
  </div>
</section>
`;
